src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs

```diff
@@ -413,6 +414,770 @@ public static class C
+            comp1.VerifyDiagnostics();
+            var ref1 = comp1.EmitToImageReference();
+
+            var source2 = @"
+using System.Reflection;
+
+[My(2 + 2, 3 + 1)]
+public class Program
+{
+    static void Main()
+    {
+        typeof(Program).GetCustomAttribute(typeof(MyAttribute));
+    }
+}
+";
+
+            var compilation = CreateCompilation(source2, references: new[] { ref1 }, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe, parseOptions: TestOptions.RegularPreview);
+            compilation.VerifyDiagnostics();
+            CompileAndVerify(compilation, expectedOutput: "2 + 2");
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void TestIncorrectParameterNameInCallerArgumentExpressionAttribute_AttributeConstructor()
+        {
+            string source = @"
+using System;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+public class MyAttribute : Attribute
+{
+    const string pp = nameof(pp);
+    public MyAttribute([CallerArgumentExpression(pp)] string arg = ""<default>"")
+    {
+        Console.WriteLine(arg);
+    }
+}
+
+[My]
+public class Program
+{
+    static void Main()
+    {
+        typeof(Program).GetCustomAttribute(typeof(MyAttribute));
+    }
+}
+";
+
+            var compilation = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe);
+            compilation.VerifyDiagnostics(
+                // (12,22): warning CS8918: The CallerArgumentExpressionAttribute applied to parameter 'arg' will have no effect. It is applied with an invalid parameter name.
+                //     static void Log([CallerArgumentExpression(pp)] string arg = "<default>")
+                Diagnostic(ErrorCode.WRN_CallerArgumentExpressionAttributeHasInvalidParameterName, "CallerArgumentExpression").WithArguments("arg").WithLocation(9, 25)
+                );
+            CompileAndVerify(compilation, expectedOutput: "<default>");
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void TestCallerArgumentWithMemberNameAttributes_AttributeConstructor()
+        {
+            string source = @"
+using System;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+public class MyAttribute : Attribute
+{
+    const string p = nameof(p);
+    public MyAttribute(int p, [CallerArgumentExpression(p)] [CallerMemberName] string arg = ""<default>"")
+    {
+        Console.WriteLine(arg);
+    }
+}
+
+[My(default(int))]
+public class Program
+{
+    static void Main()
+    {
+        typeof(Program).GetCustomAttribute(typeof(MyAttribute));
+    }
+}
+";
+
+            var compilation = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe);
+            // PROTOTYPE(caller-expr): This is inconsistent with the behavior of invocations, which doesn't show a lang version error.
```

Comment:
>PROTOTYPE(caller-expr): This is inconsistent with the behavior of invocations, which doesn't show a lang version error. [](http://example.com/codeflow?start=15&length=119)

It feels like this isn't just a language version warning. Are we certain the right default value is supplied for the parameter when the language version is "Preview"?

Reply:
@AlekseyTs Since this is an error, nothing should be supplied. But since the error is extra, and only the warning should remain, the default value should be whatever `CallerMemberName` says.

Reply:
>  Since this is an error, nothing should be supplied. 

It is an error only when the target language version is C# 9. The error is supposed to go away when the target language version is Preview

Reply:
@AlekseyTs That is how it is currently. But it shouldn't be an error in any version.

Per your recommendation on the first PR, we check language version when there is "actual binding" for it, which is not the case here.

Let me know if I'm misunderstanding something.

Reply:
> Let me know if I'm misunderstanding something.

I agree with you, the error shouldn't be there. I just find the PROTOTYPE comment somewhat incomplete. I believe the problem can be observed through a value used for the parameter in a success scenario. That raises severity of the issue in my opinion.

Reply:
@AlekseyTs This brings another question. Is the existing behavior for CallerMemberName is correct?

![image](https://user-images.githubusercontent.com/31348972/119944737-ed2ac400-bf94-11eb-98fb-6235278d2bf0.png)

It doesn't seem to work when the attribute is placed on classes. Is that by design, or a bug?

Reply:
@AlekseyTs Note that Mono compiler have a different behavior for this `CallerMemberName` case.

https://onlinegdb.com/vponLAJ2g

Reply:
> It doesn't seem to work when the attribute is placed on classes. Is that by design, or a bug?

Hard to tell, I can see this both ways. It would be interesting to know what is the behavior of the native compiler (the last version of the compiler not based on Roslyn codebase). In any case, this feels like an ortogonal issue to what we are trying to do here, feel free to open a separate issue. I think the behavior around default value shouldn't change for this scenario.

Reply:
> > It doesn't seem to work when the attribute is placed on classes. Is that by design, or a bug?
> 
> Hard to tell, I can see this both ways. It would be interesting to know what is the behavior of the native compiler (the last version of the compiler not based on Roslyn codebase). In any case, this feels like an ortogonal issue to what we are trying to do here, feel free to open a separate issue. I think the behavior around default value shouldn't change for this scenario.

➡️ https://github.com/dotnet/roslyn/issues/53757